net/ip.c

```diff
@@ -0,0 +1,343 @@
+							"inet6",
+							"unix" };
+
+static const char *ip_protocol[M0_NET_IP_PROTO_MAX] = { "tcp",
+							"verbs",
+							"stream",
+							"dgram",
+							"o2ib",
+							"lo" };
+
+/**
+ * This function convert the hostname/FQDN format to ip format.
+ * Here hostname can be FQDN or local machine hostname.
+ */
+static int m0_net_hostname_to_ip(char *hostname , char* ip,
+				 struct m0_net_ip_addr *ip_addr)
+{
+	struct hostent  *hname;
+	struct in_addr **addr;
+	int              i;
+	int              n;
+	char            *cp;
+	char             name[M0_NET_IP_STRLEN_MAX];
+
+	M0_ENTRY("Hostname = %s", (char*)hostname);
+	cp = strchr(hostname, '@');
+	if (cp == NULL)
+		return M0_ERR(-EINVAL);
+
+	n = cp - hostname;
+	memcpy(name, hostname, n);
+	name[n] = '\0';
+	if ((hname = gethostbyname(name)) == NULL) {
+		/** get the host info */
+		M0_LOG(M0_ERROR, "gethostbyname error %s", (char*)name);
+		return M0_ERR(-EPROTO);
+	}
+
+	addr = (struct in_addr **) hname->h_addr_list;
+	for(i = 0; addr[i] != NULL; i++) {
+		/** Return the first one */
+		strcpy(ip , inet_ntoa(*addr[i]));
+		n=strlen(ip);
+		M0_LOG(M0_DEBUG, "%s[%s]", (char*)name, (char*)ip);
+		if (strcmp(name, ip) == 0)
+			ip_addr->na_format = M0_NET_IP_INET_IP_FORMAT;
+		else
+			ip_addr->na_format = M0_NET_IP_INET_HOSTNAME_FORMAT;
+		return M0_RC(n);
+	}
+	return M0_ERR(-EPROTO);
+}
+
+/**
+ * This function decodes the inet format address.
+ *  The inet address format is of type
+ *    <family>:<type>:<ipaddr/hostname_FQDN>@<port>
+ *    for example: "inet:tcp:127.0.0.1@3000",
+ *                 "inet:stream:lanl.gov@23",
+ *                 "inet6:dgram:FE80::0202:B3FF:FE1E:8329@6663"
+ */
+
+static int m0_net_ip_inet_parse(const char *name, struct m0_net_ip_addr *addr)
+{
+	int          shift = 0;
+	int          f;
+	int          s;
+	char        *at;
+	char         ip[M0_NET_IP_STRLEN_MAX] = {};
+	int          n;
+	char         node[M0_NET_IP_STRLEN_MAX] = {};
+	char         port[6] = {};
+	const char  *ep_name = name;
+	uint32_t     portnum;
+
+	for (f = 0; f < ARRAY_SIZE(ip_family); ++f) {
+		if (ip_family[f] != NULL) {
+			shift = strlen(ip_family[f]);
+			if (strncmp(ep_name, ip_family[f], shift) == 0)
+				break;
+		}
+	}
+	if (ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	for (s = 0; s < ARRAY_SIZE(ip_protocol); ++s) {
+		if (ip_protocol[s] != NULL) {
+			shift = strlen(ip_protocol[s]);
+			if (strncmp(ep_name, ip_protocol[s], shift) == 0)
+				break;
+		}
+	}
+	if (ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	at = strchr(ep_name, '@');
+	if (at == NULL) {
+		return M0_ERR(-EINVAL);
+	} else {
+		at++;
+		if (at == NULL)
+			return M0_ERR(-EINVAL);
+		memcpy(port, at, (strlen(at)+1));
+		portnum = atoi(port);
+		M0_ASSERT(portnum < M0_NET_IP_PORT_MAX);
+		addr->na_port = (uint16_t)portnum;
+	}
+
+	n = m0_net_hostname_to_ip((char *)ep_name, ip, addr);
+	if (n > 0)
+		memcpy(node, ip, n);
```

Comment:
Not required. Directly use ip instead of node in below code and remove node variable.

Reply:
Done